[WIP] invert mapping b/w binop operators and the operator module #2297

asodeur opened the pull request:
This is a naive attempt at inverting the mapping between binops and the operator module to
make overloading of builtin operations easier [#2297](https://github.com/numba/numba/issues/2297). See [here](https://groups.google.com/a/continuum.io/d/msg/numba-users/zdByg1-sW8o/z5zWFvnjAAAJ) for a usecase.

While this approach seems to be feasible it runs into trouble with functionality implemented
via rewrites (like array expressions, parfor, ...). Currently most tests from `test_array_exprs.py`, `test_array_analysis.py`, `test_parfors.py`, and  `test_stencils.py` do not pass. Array expressions etc
had to be rewritten in terms of the operator module as well.

I do not understand the inner workings of the compiler sufficiently but I am slightly worried the issues
above are just the tip of an iceberg and I am creating lots of  (maybe more subtle) difference in behaviour for binops and operator module functions. Would a rewrite passes that map bewteen operator module functions and the coresponding corresponding binops such that other rewrite passes only ever get to see either binops or operator module functions make more sense?


<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

DrTodd13 commented:
From what I understand of this PR, it seems okay from the ParallelAccelerator perspective.  The vast bulk of the PR is more core Numba stuff so it is up to @sklam to decide.

One question that springs to mind is whether @sklam's original suggestion for how to do what you wanted with lower level routines would still work here.  Given that it once existed, if this change is accepted then it would seem that both techniques would need to be supported going forward in perpetuity.

asodeur commented:
Low-level extension of operator module functions should still work. However this PR would introduce some breaking changes visible to users extending builtin operators. I'll use the pair `+` and `add` (short for  `operator.add`) throughout but these points apply to any pair of builtin operator and operator module function.

Before this PR `+` and `add` are typed and implemented separately (though `add` is reusing `+`'s typing via `typing/operatordecl.py` and the implementation is calling `+` via `targets/operatorimpl.py`). After the PR there is no more explicit typing and lowering for `+`   but `+` is effectively mapped onto `add` in `ir.py`/`typeinfer.py` resp. `lowering.py`. This has the following knock-on effects (luckily only (1) is visible to users and it should be possible to avoid (3) if required)

1. All typing/lowering has to move from `+` to `add` (and move from using `@infer` to `@infer_global`). This affects all internal typing (these are the changes to `typing/*.py`) and lowering (changes to `targets/*.py`) as well as any  Numba extension. `add` can be extended via high- or low-level extension API (as before) but `+` will 'see' this extension after the PR (have to check test coverage).
2. Before this PR you could extend `+` and `add` such that `+` and `add` would behave differently. This is no longer possible. It should be possible to monkey patch `add` but with the current implementation this will not work (added TODO above)
3. Implementing the mapping of `+` to `add` I changed the `.fn` attribute in the IR representation of binops (and unary ops). This is what is causing changes to `array_analysis.py`, `array_exprs.py`,  `parfor.py`, and `stencil_parfor.py`. I am still unsure if this was a good idea. The only argument in favour probably is that this keeps a direct mapping of IR to implementation.

Another little glitch is that even after this PR `array_analysis.py` etc will not act on `add` etc. This would require `+` and `add` to have the same representation in the IR (or taking care of the operator module functions explicitly).

sklam commented:
Sorry it take me so long, this is great work and it's going to the right direction.  There're some changes that can be simplified.  For instance, I think the changes to ir.py can be actually applied at the interpreter.py level.  I will have a go and post updates here.

sklam commented:
My WIP branch: https://github.com/sklam/numba/tree/fixpr/2983

asodeur commented:
Merged https://github.com/sklam/numba/tree/fixpr/2983 and rebased on master.

I am still unsure about how to get from the operator module function in `ir.Expr.fn` to the typing template needed to retrieve the implementation in the `lowering.Lower.lower_binop`. Currently I am doing 
```python
# lowering.py, line 508
def lower_binop(self, resty, expr, op):
   if op in utils.OPERATORS_TO_BUILTINS:
      op = self.context.typing_context.resolve_value_type(op)
    ...
```
This can work if inplace operators for immutable types are explicitly lowered as in:
```python
@lower_builtin(operator.truediv, types.Integer, types.Integer)
@lower_builtin(operator.itruediv, types.Integer, types.Integer)
def int_truediv_impl(context, builder, sig, args):
   ...
```

and the reuse of copying operators for immutable types is removed from `lower_expr`: 
```python
# lowering.py, line 770
elif expr.op == 'inplace_binop':
	lty = self.typeof(expr.lhs.name)
	if lty.mutable:
		return self.lower_binop(resty, expr, expr.fn)
	else:
		# inplace operators on non-mutable types reuse the same
		# definition as the corresponding copying operators.)
		return self.lower_binop(resty, expr, expr.fn)  # used to be .immutable_fn
   ...
```
Should the lookup of the typing template rather go somewhere else than `lower_binop`? 

Andreas

sklam commented:
@asodeur 

I can restore the removed `.immutable_fn` with this change: https://github.com/sklam/numba/commit/aa88dd49e52f7f2ae7a98a60f2c4fe520313efbc .  I added a `.get_call_type` so that the "implementation" is cached properly for `.get_function`.  There is a bigger refactoring that is needed to clean this up, but that's another PR for another day.  This PR is now stable.  I will do another round of review to get it ready for merging.

sklam commented on numba/tests/test_operators.py:
```diff
@@ -379,19 +379,33 @@ class TestOperators(TestCase):
+        'bitwise_not_usecase_binary': operator.invert,
     }
 
+    # _bitwise_opnames = {
```
need cleanup

sklam commented on numba/targets/arrayobj.py:
```diff
@@ -3619,7 +3620,7 @@ def arange(start, stop, step, dtype):
             for i in range(nitems):
                 arr[i] = val
-                val += step
+                val = operator.add(val, step)  # TODO: buildin operators vs compile_internal
```
Is this change (still) necessary?

saulshanabrook commented:
Would getting `setitem`/`getitem` supported in `@overload` be best done by extending this to switch those to operators as well?

sklam commented:
> Would getting setitem/getitem supported in @overload be best done by extending this to switch those to operators as well?

Yes

sklam commented:
i'm doing a final check on this to get it merged.
